server/src/main/java/org/elasticsearch/common/geo/GeoUtils.java

```diff
@@ -459,6 +459,50 @@ public static GeoPoint parseGeoPoint(XContentParser parser, GeoPoint point, fina
         }
     }
 
+    /**
+     * Parse a precision that can be expressed as an integer or a distance measure like "1km", "10m".
+     *
+     * @param parser {@link XContentParser} to parse the value from
+     * @return int representing precision
+     */
+    public static int parsePrecision(XContentParser parser) throws IOException, ElasticsearchParseException {
+        XContentParser.Token token = parser.currentToken();
+        if (token.equals(XContentParser.Token.VALUE_NUMBER)) {
+            return XContentMapValues.nodeIntegerValue(parser.intValue());
+        } else {
+            String precision = parser.text();
+            try {
+                // we want to treat simple integer strings as precision levels, not distances
+                return XContentMapValues.nodeIntegerValue(Integer.parseInt(precision));
+            } catch (NumberFormatException e) {
+                // try to parse as a distance value
+                try {
+                    return checkPrecisionRange(GeoUtils.geoHashLevelsForPrecision(precision));
+                } catch (NumberFormatException e2) {
+                    // can happen when distance unit is unknown, in this case we simply want to know the reason
+                    throw e2;
+                } catch (IllegalArgumentException e3) {
+                    // this happens when distance too small, so precision > 12. We'd like to see the original string
+                    throw new IllegalArgumentException("precision too high [" + precision + "]", e3);
+                }
```

Comment:
should we catch separately the call to checkPrecisionRange and geoHashLevelsForPrecision?